src/AutoRest.CSharp/Mgmt/Generation/MgmtClientBaseWriter.cs

```diff
@@ -767,19 +808,39 @@ protected virtual void WriteNormalMethodBranch(MgmtRestOperation operation, IEnu
         }
         #endregion
 
+        protected bool NeedFactoryMethod(MgmtClientOperation clientOperation)
+        {
+            if (This is not Resource resource || resource.BaseResource == null)
+                return false;
+
+            return resource.BaseResource.CommonOperations.Any(operation => operation.Contains(clientOperation));
+        }
```

Comment:
Maybe instead of using this method, we could override the `WriteMethod` method in PolymorphicResourceWriter, to evaluate this bool there, and pass it through.
On those method, this new parameter could be optional and defaults to `false`.

Reply:
BTW we also need to find those places that are directly returning the concrete types and adding a cast there

Reply:
Or let them to call the corresponding Core method to keep things consistent (like our SetTags are calling Update, we should call UpdateCore if there is one, and if there is not UpdateCore, we should add a cast to keep it compile)